packages/workspace/src/browser/workspace-service.ts

```diff
@@ -88,78 +83,103 @@ export class WorkspaceService implements FrontendApplicationContribution {
         return this.onWorkspaceChangeEmitter.event;
     }
 
+    protected async onWorkspaceFileChanged(newWorkspaceFile?: FileStat): Promise<void> {
+        if (newWorkspaceFile) {
+            if (this._workspaceFile && this._workspaceFile.uri !== newWorkspaceFile.uri) {
+                this.stopWatch(this._workspaceFile.uri);
+            }
+            this._workspaceFile = newWorkspaceFile;
+        }
+        this.updateTitle();
+        if (this._workspaceFile) {
+            await this.doWatch(this._workspaceFile);
+            this.watcher.onFilesChanged(event => {
+                if (FileChangeEvent.isAffected(event, new URI(this._workspaceFile!.uri).withScheme('file'))) {
+                    this.updateWorkspace();
+                }
+            });
+        }
+    }
+
     protected async updateWorkspace(): Promise<void> {
-        await this.ensureWorkspaceInitialized();
-        await this.updateRoots();
+        const rootUri = await this.server.getMostRecentlyUsedWorkspace();
+        // `rootUri` from `recentworkspace.json` is the uri of either the workspace file or the folder to open
+        const toOpen = await this.toFileStat(rootUri);
+        await this.ensureWorkspaceInitialized(toOpen);
+        await this.updateRoots(toOpen);
         this.watchRoots();
     }
 
-    protected async ensureWorkspaceInitialized(): Promise<void> {
-        if (this._workspaceFolder) {
+    protected async ensureWorkspaceInitialized(mostRecentWorkspace: FileStat | undefined): Promise<void> {
+        if (this.opened) {
             return;
         }
-        const rootUri = await this.server.getMostRecentlyUsedWorkspace();
-        this._workspaceFolder = await this.toValidRoot(rootUri);
-        const configUri = this.getWorkspaceConfigFileUri();
-        if (configUri) {
-            this.watcher.watchFileChanges(configUri);
+        if (mostRecentWorkspace && !mostRecentWorkspace.isDirectory) {
+            this._workspaceFile = mostRecentWorkspace;
+            this.doWatch(mostRecentWorkspace);
         }
     }
 
-    protected async updateRoots(): Promise<void> {
-        this._roots = await this.computeRoots();
+    protected async updateRoots(mostRecentWorkspace: FileStat | undefined): Promise<void> {
+        if (mostRecentWorkspace) {
```

Comment:
pass `mostRecentWorkspace` down to `getWorkspaceDataFromFile` and handle it there once?